# Importing libraries and dataset

In [112]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder, OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score
from scipy.stats import zscore
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [113]:
df = pd.read_csv("heart.csv")

In [114]:
df

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

# Removing outliers

In [115]:
numerical_cols = df.select_dtypes(include=['number']).columns

In [116]:
numerical_cols

Index(['Age', 'RestingBP', 'Cholesterol', 'FastingBS', 'MaxHR', 'Oldpeak',
       'HeartDisease'],
      dtype='object')

In [117]:
zScores = zscore(df[numerical_cols])

In [118]:
zScores

Age  RestingBP  Cholesterol  FastingBS     MaxHR   Oldpeak  \
0   -1.433140   0.410909     0.825070  -0.551341  1.382928 -0.832432   
1   -0.478484   1.491752    -0.171961  -0.551341  0.754157  0.105664   
2   -1.751359  -0.129513     0.770188  -0.551341 -1.525138 -0.832432   
3   -0.584556   0.302825     0.139040  -0.551341 -1.132156  0.574711   
4    0.051881   0.951331    -0.034755  -0.551341 -0.581981 -0.832432   
..        ...        ...          ...        ...       ...       ...   
913 -0.902775  -1.210356     0.596393  -0.551341 -0.188999  0.293283   
914  1.536902   0.627078    -0.053049   1.813758  0.164684  2.357094   
915  0.370100  -0.129513    -0.620168  -0.551341 -0.857069  0.293283   
916  0.370100  -0.129513     0.340275  -0.551341  1.461525 -0.832432   
917 -1.645286   0.302825    -0.217696  -0.551341  1.422226 -0.832432   

     HeartDisease  
0       -1.113115  
1        0.898380  
2       -1.113115  
3        0.898380  
4       -1.113115  
..            ...  
913      0.898380  
914      0.898380  
915      0.898380  
916      0.898380  
917     -1.113115  

[918 rows x 7 columns]

In [119]:
noOutliers = df[(zScores < 3).all(axis=1) & (zScores > -3).all(axis=1)]

In [120]:
noOutliers

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[899 rows x 12 columns]

# converting text to numbers.

#Label Encoding for Binary Attributes

In [122]:

binary_cols= []
for col in df.columns:
    if df[col].nunique() == 2:
        binary_cols.append(col)

print("Binary Columns:", binary_cols)

Binary Columns: ['Sex', 'FastingBS', 'ExerciseAngina', 'HeartDisease']


In [60]:
labelEncoder = LabelEncoder()

In [61]:
for col in binary_cols:
  noOutliers[col]= labelEncoder.fit_transform(noOutliers[col])

<ipython-input-61-0d2e3184c05d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  noOutliers[col]= labelEncoder.fit_transform(noOutliers[col])
<ipython-input-61-0d2e3184c05d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  noOutliers[col]= labelEncoder.fit_transform(noOutliers[col])
<ipython-input-61-0d2e3184c05d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https:

In [62]:
noOutliers

Age  Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40    1           ATA        140          289          0     Normal   
1     49    0           NAP        160          180          0     Normal   
2     37    1           ATA        130          283          0         ST   
3     48    0           ASY        138          214          0     Normal   
4     54    1           NAP        150          195          0     Normal   
..   ...  ...           ...        ...          ...        ...        ...   
913   45    1            TA        110          264          0     Normal   
914   68    1           ASY        144          193          1     Normal   
915   57    1           ASY        130          131          0     Normal   
916   57    0           ATA        130          236          0        LVH   
917   38    1           NAP        138          175          0     Normal   

     MaxHR  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172               0      0.0       Up             0  
1      156               0      1.0     Flat             1  
2       98               0      0.0       Up             0  
3      108               1      1.5     Flat             1  
4      122               0      0.0       Up             0  
..     ...             ...      ...      ...           ...  
913    132               0      1.2     Flat             1  
914    141               0      3.4     Flat             1  
915    115               1      1.2     Flat             1  
916    174               0      0.0     Flat             1  
917    173               0      0.0       Up             0  

[899 rows x 12 columns]

#Using one hot encoding for categorical attributes.

In [63]:
categorical_cols = noOutliers.select_dtypes(include=['object']).columns
categorical_cols

Index(['ChestPainType', 'RestingECG', 'ST_Slope'], dtype='object')

In [64]:
noOutliers = pd.get_dummies(noOutliers, columns = categorical_cols,)

In [65]:
noOutliers

Age  Sex  RestingBP  Cholesterol  FastingBS  MaxHR  ExerciseAngina  \
0     40    1        140          289          0    172               0   
1     49    0        160          180          0    156               0   
2     37    1        130          283          0     98               0   
3     48    0        138          214          0    108               1   
4     54    1        150          195          0    122               0   
..   ...  ...        ...          ...        ...    ...             ...   
913   45    1        110          264          0    132               0   
914   68    1        144          193          1    141               0   
915   57    1        130          131          0    115               1   
916   57    0        130          236          0    174               0   
917   38    1        138          175          0    173               0   

     Oldpeak  HeartDisease  ChestPainType_ASY  ChestPainType_ATA  \
0        0.0             0                  0                  1   
1        1.0             1                  0                  0   
2        0.0             0                  0                  1   
3        1.5             1                  1                  0   
4        0.0             0                  0                  0   
..       ...           ...                ...                ...   
913      1.2             1                  0                  0   
914      3.4             1                  1                  0   
915      1.2             1                  1                  0   
916      0.0             1                  0                  1   
917      0.0             0                  0                  0   

     ChestPainType_NAP  ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  \
0                    0                 0               0                  1   
1                    1                 0               0                  1   
2                    0                 0               0                  0   
3                    0                 0               0                  1   
4                    1                 0               0                  1   
..                 ...               ...             ...                ...   
913                  0                 1               0                  1   
914                  0                 0               0                  1   
915                  0                 0               0                  1   
916                  0                 0               1                  0   
917                  1                 0               0                  1   

     RestingECG_ST  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
0                0              0              0            1  
1                0              0              1            0  
2                1              0              0            1  
3                0              0              1            0  
4                0              0              0            1  
..             ...            ...            ...          ...  
913              0              0              1            0  
914              0              0              1            0  
915              0              0              1            0  
916              0              0              1            0  
917              0              0              0            1  

[899 rows x 19 columns]

# Scaling

In [66]:
X= noOutliers.drop('HeartDisease', axis=1)
y = noOutliers['HeartDisease']
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


In [67]:
X_scaled

array([[-1.42815446,  0.515943  ,  0.46590022, ..., -0.26018448,
        -0.99888827,  1.13469459],
       [-0.47585532, -1.93819859,  1.63471366, ..., -0.26018448,
         1.00111297, -0.88129441],
       [-1.7455875 ,  0.515943  , -0.1185065 , ..., -0.26018448,
        -0.99888827,  1.13469459],
       ...,
       [ 0.3706328 ,  0.515943  , -0.1185065 , ..., -0.26018448,
         1.00111297, -0.88129441],
       [ 0.3706328 , -1.93819859, -0.1185065 , ..., -0.26018448,
         1.00111297, -0.88129441],
       [-1.63977649,  0.515943  ,  0.34901888, ..., -0.26018448,
        -0.99888827,  1.13469459]])

# Building our models

In [68]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

In [96]:
svm_model = SVC(gamma='scale',kernel = 'linear',C=1)
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)


print("Accuracy:", accuracy)
print("\nConfusion Matrix:\n", conf_matrix)
print("\nClassification Report:\n", classification_rep)

Accuracy: 0.8888888888888888

Confusion Matrix:
 [[68 18]
 [ 2 92]]

Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.79      0.87        86
           1       0.84      0.98      0.90        94

    accuracy                           0.89       180
   macro avg       0.90      0.88      0.89       180
weighted avg       0.90      0.89      0.89       180



In [94]:
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)

y_pred = logistic_model.predict(X_test)


accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)


print("Accuracy:", accuracy)
print("\nConfusion Matrix:\n", conf_matrix)
print("\nClassification Report:\n", classification_rep)

Accuracy: 0.8833333333333333

Confusion Matrix:
 [[69 17]
 [ 4 90]]

Classification Report:
               precision    recall  f1-score   support

           0       0.95      0.80      0.87        86
           1       0.84      0.96      0.90        94

    accuracy                           0.88       180
   macro avg       0.89      0.88      0.88       180
weighted avg       0.89      0.88      0.88       180



In [95]:

rf_classifier = RandomForestClassifier(max_depth= None, min_samples_leaf= 1, min_samples_split= 10, n_estimators = 200)
rf_classifier.fit(X_train, y_train)
y_pred = rf_classifier.predict(X_test)


accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)


print("Accuracy:", accuracy)
print("\nConfusion Matrix:\n", conf_matrix)
print("\nClassification Report:\n", classification_rep)

Accuracy: 0.9

Confusion Matrix:
 [[70 16]
 [ 2 92]]

Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.81      0.89        86
           1       0.85      0.98      0.91        94

    accuracy                           0.90       180
   macro avg       0.91      0.90      0.90       180
weighted avg       0.91      0.90      0.90       180



# Random Forest giving me the best accuracy


# Now, applying PCA

In [103]:

pca = PCA(n_components=len(X_scaled[0])//2)

X_pca = pca.fit_transform(X_scaled)

In [104]:
X_train, X_test, y_train, y_test = train_test_split(X_pca, y, test_size=0.2, random_state=42)

In [101]:
X_pca

array([[ 3.28879358, -0.71323178, -0.24814645, ...,  0.19770255,
         0.1756798 , -0.26459096],
       [ 0.73931996, -0.10808147, -1.632054  , ..., -0.59481057,
         0.15035689,  0.51951696],
       [ 1.93487797,  0.75186717,  1.71311368, ..., -0.67022458,
         1.17169375, -0.15644789],
       ...,
       [-1.98431191, -1.63236382, -0.55739582, ..., -0.05631287,
        -0.15951933, -0.26407742],
       [ 1.4298115 ,  2.06723082, -1.82295176, ...,  0.50109623,
         0.24688127,  2.75551035],
       [ 2.8068679 , -0.55897973,  0.09176075, ..., -0.73839005,
         0.11420657, -1.44921461]])

In [107]:
rf_classifier = RandomForestClassifier(max_depth= None, min_samples_leaf= 1, min_samples_split= 10, n_estimators = 200)
rf_classifier.fit(X_train, y_train)
y_pred = rf_classifier.predict(X_test)


accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)


print("Accuracy:", accuracy)
print("\nConfusion Matrix:\n", conf_matrix)
print("\nClassification Report:\n", classification_rep)

Accuracy: 0.8722222222222222

Confusion Matrix:
 [[69 17]
 [ 6 88]]

Classification Report:
               precision    recall  f1-score   support

           0       0.92      0.80      0.86        86
           1       0.84      0.94      0.88        94

    accuracy                           0.87       180
   macro avg       0.88      0.87      0.87       180
weighted avg       0.88      0.87      0.87       180



In [108]:
# Display the explained variance ratio after PCA
print("\nExplained Variance Ratio after PCA:")
print(pca.explained_variance_ratio_)


Explained Variance Ratio after PCA:
[0.21299289 0.10306303 0.08710183 0.07727703 0.0717471  0.06445626
 0.0628358  0.0567701  0.05143852]


In [110]:
import numpy as np
cumulative_variance = np.cumsum(pca.explained_variance_ratio_)
print("Cumulative Explained Variance:", cumulative_variance)

Cumulative Explained Variance: [0.21299289 0.31605592 0.40315775 0.48043478 0.55218189 0.61663814
 0.67947394 0.73624404 0.78768256]


# after PCA, approximately 78.8% of the original information is retained by our data.


